File: Python/ds/BinarySearchTree.py
# Implementation of Binary Search Tree


class BSTNode(object):
    # A Node of the Binary Search Tree 
    def __init__(self, val=None, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def insertNode(root, val):
    # Function to insert a node to a BST

    # if the given BST is empty
    if root == None:
        return BSTNode(val)

    else:
        # if the value to insert already exists in the tree
        if root.val == val:
            return root

        # if the value to insert should be inserted to the left subtree
        if val < root.val:
            root.left = insertNode(root.left, val)
        
        # if the value to insert should be inserted to the right subtree
        else:
            root.right = insertNode(root.right, val)

    return root

    
def searchNode(root, val):
    # Function to search a node in a BST

    # if the BST is empty or if the node doesn't exist
    if root == None:
        return False

    # if the value exists in the BST
    if root.val == val:
        return True

    # if the value might exist in the right subtree    
    elif val > root.val:
        return searchNode(root.right, val)

    # if the value might exist in the left subtree
    else:
        return searchNode(root.left, val)


def inorderSuccessor(node):
    # Function to find the inorder successor of a given node

    curr = node
    # Loop through the BST until a left most node is reached
    while (curr.left != None):
        curr = curr.left
    return curr


def deleteNode(root, val):
    # Function to delete a node in a BST

    # if the BST is empty
    if root == None:
        return root
    
    # if the node to delete might exist in the left subtree
    elif root.val > val:
        root.left = deleteNode(root.left, val)

    # if the node to delete might exist in the right subtree
    elif root.val < val:
        root.right = deleteNode(root.right, val)

    # if the node is found
    else:

        # if the node has one child or no children
        if root.right == None:
            tempNode = root.left
            root = None
            return tempNode

        elif root.left == None:
            tempNode = root.right
            root = None
            return tempNode 

        # if the node has two children
        tempNode = inorderSuccessor(root.right)
        root.val = tempNode.val
        root.right = deleteNode(root.right, tempNode.val)

    return root


def inorder(root):
    # Function to print the nodes in the BST inorder
    
    # if the BST is not empty
    if root:
        # visit the left subtree (or leaf)
        inorder(root.left)
        # visit the parent
        print(root.val, end = ' ')
        # visit the right subtree (or leaf)
        inorder(root.right)
    
    
# A menu based implementation
if __name__ == '__main__':
    bst = None
    op = 1

    while (op != 5):
        print()
        print("1: Insert")
        print("2: Search")
        print("3: Delete")
        print("4: Print (Inorder)")
        print("5: Exit")
        print("Enter option: ", end = "")
        
        op = int(input())
        
        if op == 1:
            print("Enter value to insert: ", end = "")
            val = int(input())
            bst = insertNode(bst, val)
            print()
        
        elif op == 2:
            print("Enter element to search for: ", end = "")
            val = int(input())
            if searchNode(bst, val):
                print("The element is present!")
            else:
                print("The element is not present")
            print()
        
        elif op == 3:
            print("Enter the value to delete: ", end = "")
            val = int(input())
            bst = deleteNode(bst, val)
            print()
        
        elif op == 4:
            if bst:
                inorder(bst)
            else:
                print("Empty BST!")
            print()

File: Java/ds/BinarySearchTree.java

public class BinarySearchTree {
	
	
	
	class Node{
		int data;
		Node left, right;
		
		Node(int data){
			this.data = data;
			this.left = null;
			this.right = null;
		}
		
		
	}
	
	/*
	 * Inserting a new node to BST with recursion 
	 * @param Node (root node to tree)
	 * @param value to insert
	 */
	public Node insert(Node root, int value) {
		
		if(root==null)
			root = new Node(value);
		
		else {
			
				if(root.data >= value)
					root.left = insert(root.left, value);
				else
					root.right = insert(root.right, value);
			
		}
		
		return root;	
	}
	
	/*
	 * To findPath for given key recursively 
	 * @param Node root node
	 * @param key to find
	 * return String path to node
	 */
	public String findPath(Node root, int key) {
		
		
		if(root!=null) {
			
			String seq = String.valueOf(root.data);
			
			if(key == root.data)
				return seq;
			
			else if(key< root.data) {
				String returnValue = findPath(root.left, key);
				return returnValue.equals("NF")? returnValue : seq+"-->"+returnValue;
			}
			
			else {
				String returnValue = findPath(root.right, key);
				return returnValue.equals("NF")? returnValue : seq+"-->"+returnValue;			}
		}
		
		/*
		 * Return NF if not found
		 */
		return "NF";
	}
	
	
	/*
	 * Find Node with key
	 * @param Node root node
	 * @param key to find
	 * return Node which contains key
	 */
	public Node findNode(Node root, int key) {
		
		if(root!=null) {
			
			//String seq = String.valueOf(root.data);
			
			if(key == root.data)
				return root;
			
			else if(key< root.data) {
				return findNode(root.left,key);
			}
			
			else {
				return findNode(root.right,key);
			}
		}
		
		
		return null;
	}
	
	
	/*
	 * Delete given value from BST
	 * @param Node root node
	 * @param key to delete
	 */
	public Node delete(Node root, int value) {
		
			
		if(root!=null) {
			
			//if key found
			if(root.data == value){
				
				
				if(root.left == null && root.right==null)
					//delete leaf node
					return null;
				
				
				else if(root.left != null && root.right!=null) {
					
					/*
					 * Delete node with 2 child with below steps
					 * find inorder successor of node
					 * replace value in inorder successor to current node
					 * delete inorder successor and assign it's child to right of curr node
					 */
					Integer inorderSuc = minInSubtree(root.right);
					root.data = inorderSuc;
					root.right = delete(root.right, inorderSuc);
					
				}
				else {
					//delete node with one child	
					if(root.left == null)
						return root.right;
					
					else
						return root.left;
					
				}
				
				
			}
			
			//search in left subtree
			else if(root.data > value)
				root.left = delete(root.left, value);
			//search in right subtree
			else
				root.right = delete(root.right, value);
			
			
		}
			
			
		return root;
	}
	
	
	/*
	 * Get min value of given sub tree
	 * @param Node root node of subtree
	 * return integer (min value in sub tree
	 */
	public Integer minInSubtree(Node currNode) {
		
		if(currNode.left == null)
			return currNode.data;
		else
			return minInSubtree(currNode.left);
		
		
	}
	
	
	/*
	 * Print data of current node, data in left node and right node
	 */
	public void printNodeLeftRight(Node node) {
		
		System.out.println("Node->data:- "+node.data);
		if(node.left == null)
			System.out.println("Node->left:- "+ null);
		else
			System.out.println("Node->left:- "+ node.left.data);
		
		if(node.right == null)
			System.out.println("Node->right:- "+ null);
		else
			System.out.println("Node->right:- "+ node.right.data);
		
		
	}
	
	
	/*
	 * Print in order traversal of given tree
	 */
	public void inorderTraversal(Node root) {
		
		
		if(root!=null) {
			
			inorderTraversal(root.left);
			System.out.print(root.data+"\t");
			inorderTraversal(root.right);
		}
		
		
		
	}
	
	public static void main(String[] args) {
		
		
		BinarySearchTree bst = new BinarySearchTree();
		
		Node root = null;
		
		root = bst.insert(root, 10);
		root = bst.insert(root, 20);
		root = bst.insert(root, 5);
		root = bst.insert(root, 1);
		root = bst.insert(root, 9);
		
		System.out.println("##################### PATH FOR EACH ####################");
		
		String seq;
		seq = bst.findPath(root, 10);
		System.out.println("seq for "+10+" is "+seq);
		seq = bst.findPath(root, 20);
		System.out.println("seq for "+20+" is "+seq);
		seq = bst.findPath(root, 5);
		System.out.println("seq for "+5+" is "+seq);
		seq = bst.findPath(root, 1);
		System.out.println("seq for "+1+" is "+seq);
		seq = bst.findPath(root, 9);
		System.out.println("seq for "+9+" is "+seq);

		
		bst.inorderTraversal(root);
		System.out.println("\n");
		System.out.println("\n##################### INORDER FOR AFTER EACH DELETION ####################");
		
		System.out.println("deleting :- 9");
		root = bst.delete(root, 9);
		bst.inorderTraversal(root);
		System.out.println("\n");
		
		System.out.println("deleting :- 5");
		root = bst.delete(root, 5);
		bst.inorderTraversal(root);
		System.out.println("\n");
		
		System.out.println("deleting :- 10");
		root = bst.delete(root, 10);
		bst.inorderTraversal(root);
		System.out.println("\n");
		
		
		
		System.out.println("\n##################### PATH FOR  EACH AFTER DELETION ####################");
		seq = bst.findPath(root, 10);
		System.out.println("seq for "+10+" is "+seq);
		seq = bst.findPath(root, 20);
		System.out.println("seq for "+20+" is "+seq);
		seq = bst.findPath(root, 5);
		System.out.println("seq for "+5+" is "+seq);
		seq = bst.findPath(root, 1);
		System.out.println("seq for "+1+" is "+seq);
		seq = bst.findPath(root, 9);
		System.out.println("seq for "+9+" is "+seq);
		
	
	}

}
